### PR TITLE
enhancement(datadog_events sink): Adhere to instrumentation spec

### PR DESCRIPTION
**Pull request by neuronull**
Closes: #14232

Epic: https://github.com/vectordotdev/vector/issues/13995

- Add `ComponentEventsDropped` emission
- Use `assert_sink_compliance` and `assert_sink_error`

Note- I believe CI will fail for this because of the double emission bug I introduced, but once https://github.com/vectordotdev/vector/pull/14356  is in, I will merge that in here.

### CODE REVIEW

File: src/sinks/datadog/events/mod.rs [hunk 1]
```diff
@@ -7,3 +7,5 @@ pub mod sink;
 mod tests;
 
 pub use self::config::DatadogEventsConfig;
+
+const NAME: &str = "datadog_events";
```
Comment (spencergilbert) [inline src/sinks/datadog/events/mod.rs:11]: A name should be macro'd as this is a named component, I don't believe we need this anymore.

**Review by spencergilbert — APPROVED** [review]

File: src/sinks/datadog/events/mod.rs (hunk 1 above)
Comment (neuronull) [reply src/sinks/datadog/events/mod.rs:11]: Oh neat. :thinkies: do you have an example or the name of the macro to call ?

Comment (spencergilbert) [reply src/sinks/datadog/events/mod.rs:11]: Discussed offline, but via `NamedComponent`

Comment (neuronull) [reply src/sinks/datadog/events/mod.rs:11]: Awesome, thanks! :tada:

**Review by spencergilbert — APPROVED** [review]

**Pull request merged**